sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -655,6 +655,24 @@ object TypeCoercion {
       case NaNvl(l, r) if l.dataType == FloatType && r.dataType == DoubleType =>
         NaNvl(Cast(l, DoubleType), r)
       case NaNvl(l, r) if r.dataType == NullType => NaNvl(l, Cast(r, l.dataType))
+
+      case a @ ApproximatePercentile(child, pe @ NumericType(), ae @ IntegralType(), _, _) =>
+        val newChild = if (child.dataType == NullType) Cast(child, DoubleType) else child
+        val newPe = if (pe.dataType == DoubleType) pe else Cast(pe, DoubleType)
+        val newAe = if (ae.dataType == LongType) ae else Cast(ae, LongType)
+        a.copy(child = newChild, percentageExpression = newPe, accuracyExpression = newAe)
+      case a @ ApproximatePercentile(child, pe, ae @ IntegralType(), _, _)
+          if ArrayType.acceptsType(pe.dataType) &&
+            pe.dataType.asInstanceOf[ArrayType].elementType.isInstanceOf[NumericType] =>
+        val newChild = if (child.dataType == NullType) Cast(child, DoubleType) else child
+        val ArrayType(elementType, containsNull) = pe.dataType
+        val newPe = if (elementType == DoubleType) {
+          pe
+        } else {
+          Cast(pe, ArrayType(DoubleType, containsNull))
+        }
+        val newAe = if (ae.dataType == LongType) ae else Cast(ae, LongType)
```

Comment:
After a second thought, I think we don't need to cast to long. There is a better way to evaluate `accuracyExpression`
```
accuracyExpression.eval().asInstanceOf[Number].longValue
```

Reply:
ah.. I can change it back to `accuracyExpression.eval().asInstanceOf[Number].longValue`